control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+ * its serialized representation. The field may be synthesized, so there may be
+ * no corresponding field at the IR level.
+ */
+struct HeaderField {
+  // The header instance owning this field. This is a fully qualified external
+  // name of the sort stored in a PISymbolTable.
+  const cstring instance;
+
+  // The fully qualified external name of this field. This is *also* fully
+  // qualified - that is, this *includes* @instance.
+  const cstring name;
+
+  const size_t offset;  // This field's offset (really index) within the header.
+  const uint32_t bitwidth;
+  const boost::optional<pi_p4_id_t> id;  // Any '@id' P4 annotation for this field.
+  const IR::IAnnotated* annotations;  // If non-null, this field's annotations.
+
+  /// @return a HeaderField constructed from an instance @path and an @fieldName.
+  static HeaderField* from(const HeaderFieldPath* path, cstring fieldName,
+                           size_t offset, uint32_t bitwidth,
+                           boost::optional<uint32_t> id,
+                           const IR::IAnnotated* annotations) {
+    return new HeaderField{path->serialize(), path->append(fieldName)->serialize(),
+                           offset, bitwidth, id, annotations};
+  }
+
+  /// @return a HeaderField which has no corresponding IR field. Prefer this
+  /// method over from() so synthesized fields are documented at the callsite.
+  static HeaderField* synthesize(const HeaderFieldPath* path, cstring fieldName,
+                                 size_t offset, uint32_t bitwidth) {
+    return from(path, fieldName, offset, bitwidth, boost::none, nullptr);
+  }
+};
+
+/// @return the value of any P4 '@id' annotation @declaration may have. The name
+/// 'externalId' is in analogy with externalName().
+static boost::optional<pi_p4_id_t>
+externalId(const IR::IDeclaration* declaration) {
+  CHECK_NULL(declaration);
+  if (!declaration->is<IR::IAnnotated>()) {
+    return boost::none;  // Assign an id later; see below.
+  }
+
+  // If the user specified an @id annotation, use that.
+  auto annotations = declaration->to<IR::IAnnotated>()->getAnnotations();
+  auto idAnnotation = annotations->getSingle("id");
+  if (idAnnotation != nullptr) {
+    if (idAnnotation->expr.size() != 1) {
+      ::error("@id should be an integer for declaration %1%", declaration);
+      return boost::none;
+    }
+
+    auto idConstant = idAnnotation->expr[0]->to<IR::Constant>();
+    if (idConstant == nullptr || !idConstant->fitsInt()) {
+      ::error("@id should be an integer for declaration %1%", declaration);
+      return boost::none;
+    }
+
+    const uint32_t id = idConstant->value.get_si();
+    return id;
+  }
+
+  // The user didn't assign an id.
+  return boost::none;
+}
+
+/// The global symbols which are exposed to PI. Non-global symbols include
+/// action parameters and header instance fields.
+enum class PISymbolType {
+  ACTION,
```

Comment:
This is a PI symbol table, and PI does not really have scopes - although the names we generate take into account the hierarchy that exists at the P4 level, the PI namespace is currently flat. The `HeaderFieldPath` and `HeaderField` objects do keep that scope information alive until the field names are serialized into their final PI representation, though.